docs/adr/001_encoder.md

```diff
@@ -0,0 +1,48 @@
+# 001 - Encoder used for Capturing Video
+
+## Date
+
+2023-01-26
+
+## Status
+
+Decision
+
+## Category
+
+Architecture
+
+## Authors
+
+@roamingthings, @mryel00
+
+## References
+
+[The Picamera2 Library Documentation](https://datasheets.raspberrypi.com/camera/picamera2-manual.pdf)
+
+## Context
+
+The Picamera2 library contains different encoders to capture video.
+
+We want to provide an mjpeg video stream and single still images (snapshots).
+
+This software aims at systems that will run additional tasks like 3D printers running Klipper,
+Mainsail etc. In addition we assume that only few clients will access the stream simultaneously.
+
+## Options
+
+1. Use `JpegEncoder` a multi-threaded software JPEG encoder
+2. Use `MJPEGEncoder` an MJPEG encoder using the Raspberry Pi’s hardware
+
+## Decision
+
+We will use the `MJPEGEncoder` that is using the Raspberry Pi's hardware.
+
+## Consequences
+
+* Following the documentation this encoder will consume less CPU than the software encoder
+* Too many simultaneous streams will put more load on the CPU than the software encoder.
```

Comment:
Did you test it? In my testings I get to like 15% (MJPEG) with 12 streams vs ~100%(Jpeg) with no stream at all.

Reply:
I have to correct me and have to add some more data. The MJPEG is ~13% and Jpeg is ~40% with 1080p 10fps 12 streams (6 chrome + 6 firefox).
At 1080p 50fps I get about ~50% (MJPEG) vs ~350% (Jpeg) with 12 streams (6 chrome + 6 firefox).
Without a stream open and 1080p 50fps I get about ~15% (MJPEG) vs 270% (Jpeg).
Conclusion MJPEG is always better as it doesn't call a C library to use software encoding like Jpeg.

Reply:
Thank you for the benchmarking.
I didn't test it myself yet. This is what I understood from the discussion in #8. Seems as if I was wrong :she

Reply:
I just realized your misassumption myself and want to explain it more in detail what I meant there. I also made a mistake there.
The GIL in Python makes that only one thread might execute at the same time. If you now get a lot of requests on your server, you get more threads to handle as you have update every single one of them.
So the encoder might not get the GIL to write the next frame in to the buffer. So every single stream instance will only update to the same frame if the encoder won't be called between them. In worst case every stream could get updated to the same frame mulitple times.
With multi-threading encoding the times that the encoder gets called is increasing that not all stream instances get updated to the same frame. Also this only makes sense if the GIL is getting released by the encoder itself in the encode function to call an extern library that could take a bit longer to encode.
As we don't need the multi-threading behaviour there is no need for the Jpeg encoder and **_even then we should write our own encoder as the Jpeg encoder is not optimized for performance_**.
I can't guarantee the accuracy of everything I wrote there but it should explain my previous comment a bit better.